### DISCUSSION TITLE
Error that I can't resolve

### DISCUSSION THREAD

**Discussion by fraygon** [jpcornil-git/HA-mcp23017#27] [Q&A]
I've been receiving this message for some time and I can't understand or resolve it. 
My H.A. it is updated to the latest version, runs on an rpi4. 
Your integration has always worked on three 23017 chips although sometimes I lose some entities and have to reboot.

> Questo errore ha avuto origine da un'integrazione personalizzata.

Registratore: root
Fonte: custom_components/mcp23017/__init__.py:258
Integrazione: MCP23017 Digital I/O Expander (documentazione, problemi)
Prima occorrenza: 15:28:57 (1 occorrenze)
Ultima registrazione: 15:28:57

Uncaught thread exception
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/threading.py", line 1073, in _bootstrap_inner
    self.run()
  File "/config/custom_components/mcp23017/__init__.py", line 385, in run
    input_state = input_state & 0x00FF | (self[GPIOB] << 8)
                                          ~~~~^^^^^^^
  File "/config/custom_components/mcp23017/__init__.py", line 258, in __getitem__
    data = self._bus.read_byte_data(self._address, register)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/smbus2/smbus2.py", line 433, in read_byte_data
    ioctl(self.fd, I2C_SMBUS, msg)
OSError: [Errno 121] Remote I/O error >

Thanks for the help.

**Comment by jpcornil-git** [comment]
It looks like a HW issue with your i2c bus or address. Did you test access  to start with (i2cdetect as explained [in a previous discussion](https://github.com/jpcornil-git/HA-mcp23017/issues/7) to confirm device presence and then i2set/get to test access)

**Comment by fraygon** [comment]
Thanks for the quick reply.
Compared to the previous discussion, the integration now works well. The "error" warning appears when I do a reboot.
I'm trying to resolve various "errors" and "warnings" that I have at startup and during operation.
This is what I read based on your suggestion.
![Immagine 2024-06-25 115344](https://github.com/jpcornil-git/HA-mcp23017/assets/69857110/c4921e66-d02e-4726-a914-f1943ddee125)

**Comment by jpcornil-git** [comment]
What shell is this ? HAOS ssh addon I guess ?

**Comment by fraygon** [comment]
YES. Sorry I didn't specify it.

**Comment by jpcornil-git** [comment]
The ssh addon runs in a dedicated docker container (like any addon or homeassistant) on top of HAOS and has limited rights and visibility on the whole system (/config directory, ha tools, ...)

In order to access the homeassistant container, you need first ssh access to HAOS, see 
https://developers.home-assistant.io/docs/operating-system/debugging/#enabling-ssh-access-to-the-host 

Once logged in HAOS, you can see all docker containers by running a `docker ps` command (there should be one called homeassistant) and start a shell within one of them, e.g. homeassistant
```
docker exec -it homeassistant /bin/bash
```
Within that container, you should  now have access rights to /dev/i2c-xx and be able to run above i2cdetect command

**Comment by fraygon** [comment]
I tried, up to a certain point, with the procedure you suggested but I stopped for fear of damaging my installation (I'm not a programmer but just a circuit design technician).
Thank you for the time you dedicated to me.
I'm keeping the error and hoping that everything will continue to work anyway.

**Comment by jpcornil-git** [comment]
Thanks, risks are very limited here as you are only enabling ssh to the host and starting a shell within a container and observe internals.
Anyway, your call :-) but your issue looks related to hardware somehow (i2c bus availability, signal integrity, ...) 
Cheers,
jpc
